pkg/collector/reconcile/horizontalpodautoscaler.go

```diff
@@ -0,0 +1,144 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reconcile
+
+import (
+	"context"
+	"fmt"
+
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/open-telemetry/opentelemetry-operator/pkg/collector"
+)
+
+// +kubebuilder:rbac:groups=autoscaling,resources=horizontalpodautoscalers,verbs=get;list;watch;create;update;patch;delete
+
+// HorizontalPodAutoscaler reconciles HorizontalPodAutoscalers if autoscale is true and replicas is nil.
+func HorizontalPodAutoscalers(ctx context.Context, params Params) error {
+	desired := []autoscalingv1.HorizontalPodAutoscaler{}
+
+	// check if autoscale mode is on
+	if isHPARequired(params) {
+		desired = append(desired, collector.HorizontalPodAutoscaler(params.Config, params.Log, params.Instance))
+	}
+
+	// first, handle the create/update parts
+	if err := expectedHorizontalPodAutoscalers(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the expected horizontal pod autoscalers: %w", err)
+	}
+
+	// then, delete the extra objects
+	if err := deleteHorizontalPodAutoscalers(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the horizontal pod autoscalers: %w", err)
+	}
+
+	return nil
+}
+
+func isHPARequired(params Params) bool {
+	return params.Instance.Spec.Autoscale != nil && *params.Instance.Spec.Autoscale && params.Instance.Spec.Replicas == nil
+}
+
+func expectedHorizontalPodAutoscalers(ctx context.Context, params Params, expected []autoscalingv1.HorizontalPodAutoscaler) error {
+	for _, obj := range expected {
+		desired := obj
+
+		if err := controllerutil.SetControllerReference(&params.Instance, &desired, params.Scheme); err != nil {
+			return fmt.Errorf("failed to set controller reference: %w", err)
+		}
+
+		existing := &autoscalingv1.HorizontalPodAutoscaler{}
+		nns := types.NamespacedName{Namespace: desired.Namespace, Name: desired.Name}
+		err := params.Client.Get(ctx, nns, existing)
+		if err != nil && k8serrors.IsNotFound(err) {
+			if err := params.Client.Create(ctx, &desired); err != nil {
+				return fmt.Errorf("failed to create: %w", err)
+			}
+			params.Log.V(2).Info("created", "hpa.name", desired.Name, "hpa.namespace", desired.Namespace)
+			continue
+		} else if err != nil {
```

Comment:
This construct can be made easier to read e.g. 

```golang
if err != nil {
 if isNotFound(err) {
 // create
 } else {
  return fmt.Errorf("failed to get %w", err)
 }
}
```

or the following if `isNotFound` can be called with `nil` (I think it can)

```golang
if isNotFound(err){
} else err != nil {
}
```

Reply:
updated